lms/models/lti_event.py

```diff
@@ -0,0 +1,94 @@
+from enum import Enum
+from functools import lru_cache
+
+import sqlalchemy as sa
+from sqlalchemy.dialects.postgresql import JSONB
+from sqlalchemy.ext.mutable import MutableDict
+from sqlalchemy import insert
+
+from lms.db import BASE
+
+
+class LTIEventType(BASE):
+    __tablename__ = "lti_event_type"
+
+    class Type(str, Enum):
+        LAUNCH = "launch"
+        DEEP_LINKING = "deep_linking"
+
+    id = sa.Column(sa.Integer, autoincrement=True, primary_key=True)
+    type = sa.Column(
+        "type",
+        sa.Enum(
+            Type,
+            # In order to maintain maximum flexibility we will only enforce the
+            # type on the Python side, and leave the Postgres side open as a plain
+            # VARCHAR
+            native_enum=False,
+            create_constraint=False,
+            validate_strings=True,
+            # Without a length SQLAlchemy will constrain it to the longest value
+            # we happen to have right now, which could limit us in future
+            length=64,
+            # Use the string values, not the keys to persist the values
+            values_callable=lambda obj: [item.value for item in obj],
+        ),
+        nullable=False,
+        unique=False,
+    )
+
+
+class LTIEvent(BASE):
+    """Track each LTI launch."""
+
+    __tablename__ = "lti_event"
+
+    id = sa.Column(sa.Integer, autoincrement=True, primary_key=True)
+
+    created = sa.Column(
```

Comment:
Not using the mixin, we only care about `created`.

Reply:
This is such a nit picky thing, but when I was thinking about this I thought this might want to be something like `timestamp` instead.

There's a bit of confusion when we use `created` and `updated` in our DB. For example in `assignment` table we have a `created` row, but it doesn't record when the assignment was created, but when the row in our DB was created. The assignment could have been created a long time before, and then deep linking setup etc. If we  were table to track the actual create date too, we'd have a bit of a naming problem on our hands.

It's conceivable (for the sake of argument, rather than because I think it'll happen) that we might infer an event happened at a different time and then store that time. For example we scan some activity and detect events and save them with different timestamps (other than now). So the DB row might get created now, but the event might have happened yesterday because we detected it during a schedule task or something.

This is only a semantic difference, but I think our `created` and `updated` columns are basically _metadata_ about when we stored information, but I think the event `timestamp` is _primary data_ about the event.